### ISSUE TITLE
Optimize `Channel.select_impl`

### ISSUE THREAD

**Issue by straight-shoota** [crystal-lang/crystal#12756]
`Channel.select_impl` allocates an array for all action locks, even if there is only a single one (https://github.com/crystal-lang/crystal/pull/12694#issuecomment-1296196003). This can be avoided.

Most typically it should only be a handful of actions (< 10 or), so we could consider using StaticArray for stack allocation.

**Comment by straight-shoota** [comment]
This could be a patch for this, but I'd like to get it a bit more concise and it's not sufficiently tested. The spec suite has only one example with duplicates and removing the uniqueness filter doesn't break it.
```diff
--- i/src/channel.cr
+++ w/src/channel.cr
@@ -458,12 +458,30 @@ class Channel(T)
   end

   private def self.select_impl(ops : Indexable(SelectAction), non_blocking)
+    # Avoid heap allocation for typical small ops sizes (< 10 actions)
+    ops_buffer = uninitialized StaticArray(typeof(ops.each{|o| break o}.not_nil!), 10)
+    if ops.size <= ops_buffer.size
+      ops.each_with_index do |op, i|
+        ops_buffer.to_unsafe[i] = op
+      end
+      ops_locks = ops_buffer.to_slice[0, ops.size]
+    else
+      ops_locks = ops.to_a
+    end
+
     # Sort the operations by the channel they contain
     # This is to avoid deadlocks between concurrent `select` calls
-    ops_locks = ops
-      .to_a
-      .uniq!(&.lock_object_id)
-      .sort_by!(&.lock_object_id)
+    ops_locks.unstable_sort_by!(&.lock_object_id)
+    duplicates = 0
+    ops_locks.each_with_index do |lock, i|
+      next if i == 0
+      if lock.lock_object_id == ops_locks[i - duplicates - 1].lock_object_id
+        duplicates += 1
+      else
+        ops_locks[i - duplicates] = lock
+      end
+    end
+    ops_locks = ops_locks[0, ops_locks.size - duplicates]

     ops_locks.each &.lock
```

We need two separate features for implementing this which I think could be extracted because they are useful on their own:

* Materialize an `Enumerable` into a stack-allocated `Slice`
* Remove duplicates from `Slice`.

**Comment by asterite** [comment]
I don't think this is the way to go. The compiler rewrites a `select` into something else. Let the compiler allocate the static array with exactly the size needed, get a slice to it and pass it to `select_impl`. Super easy, and it works for all cases.

**Comment by straight-shoota** [comment]
A bigger refactoring is suggested in https://github.com/crystal-lang/crystal/pull/12694#issuecomment-1296275035

> Also it seems an Indexable is passed to select_impl. In the actual implementation that's a tuple. But that's not good if we need to sort the values. So I would change the implementation to use a Static Array and pass a slice to the method. Then the slice is mutable so the data can be sorted in-place. And maybe then the data can also be structs to avoid memory allocations. But I see some pointers being passed around, so it's a tricky refactor. But even though it'll take some time to get it right, it's probably the best thing to do.

**Comment by straight-shoota** [comment]
It shouldn't take much to switch the compiler to emit `StaticArray`/`Slice` instead of `Tuple`. That should be fully compatible with the existing stdlib code because it receives `Indexable`. So that's good. I'll submit a patch for that shortly.

On its own, this doesn't change anything though. In order to improve performance we have to avoid heap allocations in `Channel.select_impl`. There is definitely an array allocation for `ops_locks`. Additionally, if there are more then `Array::SMALL_ARRAY_SIZE` ops, `uniq!` would also allocate a hash.

This array is basically an ordered list of distinct lock objects. It prevents deadlocks from trying to lock the same lock twice (uniqueness) as well as deadlocks between concurrent calls (deduplication).
The implementation looks like this:

https://github.com/crystal-lang/crystal/blob/d40b44b66df1bea01e8deb4289a65d7aea362501/src/channel.cr#L423-L459

The uniqueness can easily be avoided when the list is sorted anyways. Identical lock_ids are ordered in sequence and can be skipped on repeat.

If the sorting would apply to the original slice, it disrupts the order of actions so that their index wouldn't correlate to the index of the `when` branch anymore.
We could prevent that by encoding the index into the `SelectAction`.
Alternatively, `select_impl` could receive `StaticArray` and use another instance of that type as a stack-allocated duplicate for sorting. This should work fine with the proposed compiler change (which goes to `StaticArray`).

**Comment by straight-shoota** [comment]
We could take inspiration from the `select` implementation in Golang: It uses a stack-allocated array of indices into the operations array and orders these indices.
The buffer is apparently provided by the runtime. But its a static array capped at 65_535 max (`UInt16::MAX`). That should certainly be big enough for any select statement. Each index is a `UInt16`, so it takes up to 128kiB. When the number of operations is known at compile time, the static array can be smaller of course.

https://github.com/golang/go/blob/527610763b882b56c52d77acf1b75c2f4233c973/src/runtime/select.go#L188-L222

I believe this could be a good solution.
